### PR TITLE
Holoparasite rework: build-a-bear edition

### LINKED ISSUE #11847: Holoparasite rework: build-a-stand edition
[Title]: Holoparasite rework: build-a-stand edition

[Implementation]: So basically, there's no more "stand types". Instead, each stand has stuff like "major ability", "minor ability", "defense", "offense", "range", et cetera. And you only get X amount of points to put into it total. So you can make the ULTIMATE DEFENSIVE STAND, or "literally just a stand-rush spammer", or "it sets people on fire when it punches them" or whatever. So you could get all sorts of gimmicky stands (or just use a preset that someone will inevitably make and post on the KB discord)

also a stand arrow which just gives you a randomized stand but probably kills you

and also actual stand rushes using `mousedown`

[Resources]: i dunno, someone who knows HTML, maybe? i could probably do this with tgui but dwarf will hate me for it.

thanks to @Manly-Dwarf for the idea

@HippieStation/designers

### PR DESCRIPTION
**Pull request by ghost**
:cl:
add: Guardians/holoparasites are no longer seperated into types
add: Instead, you get to "build" one, with a limited amount of points, selecting different stats and abilities
add: Will probably be very unbalanced the first few times.
del: Goodbye lightning holopara, nobody will ever miss you.
/:cl:

closes https://github.com/HippieStation/HippieStation/issues/11847

### CODE REVIEW

**Comment by NautilusViper** [comment]
this PR got me breathin something fierce

+1 and it isn't even done yet

**Comment by NautilusViper** [comment]
healing ability is broke dick as fuck with max persistence, it also shouldn't be able to heal itself

**Comment by ghost** [comment]
Yeah but your offensive will suck though

**Comment by mikolaj007** [comment]
oh man, sounds awesome +1 instantly

**Comment by Manly-Dwarf** [comment]
holy shit do not have self-healing

**Comment by ghost** [comment]
lmao @ secret sauce causing travis to randomly fail

**Comment by ghost** [comment]
#### Test Merge Deployed

##### Server Instance
HippieStation

##### Revision
Origin: 9bf2c62463c3bd5b4f5062016a66dff8da27d223
Pull Request: 8b7d666397944d13b8b5e3f64b2800391dcf17c9
Server: 1a4c29f1d287e617243c0030809ebb3399dfaf94

##### Comment
Travis CI has failed but we're testing it any way

**Comment by NautilusViper** [comment]
by the way, 'persistence' doesn't really reflect what the stat does, maybe change it to "potential" to reflect actual stand stats, even if potential doesn't really mean the same thing it DOES feel closer to me
oh, and defense should be called durability, if we're taking that route

**Comment by ExpAntique** [comment]
+1 if you add that wierd hexagon graph thing that jojo uses to show a stand’s stats

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 9bf2c62463c3bd5b4f5062016a66dff8da27d223
Pull Request: 77d51e624fb76e0f4fe65fe88ebf558191f4d361
Server: ccfb70a20c712d40a9d2a14d17c13b8c337b83dc

##### Comment
R.I.P. Gold Experience Requiem, finally found its weakness.

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 72e72e24825f69dc8cfa86a36db3a9577043fe3e
Pull Request: 77d51e624fb76e0f4fe65fe88ebf558191f4d361
Server: 4c659a75acdb4032c0951b768d9d2dd50e78b2c9

##### Comment
why do I need to update these testmerges every single time I just want to compile

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 72e72e24825f69dc8cfa86a36db3a9577043fe3e
Pull Request: b64ba336d30d069d0a37712b1c4291a58ea5a6ec
Server: 472ce7fa8d70296e12e68c175967b9cf10a42994

##### Comment
guess we're doin this again

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 4705fd9e7ae76fea77ac58bc62456e73152df7f4
Pull Request: 595828acb89adf8d820c3e2bda6d10ec12f19c0e
Server: ae9a7a0a76f9d9fece3506fe2efa31e37e6a80f9

##### Comment
misspelled "means" there, steamp0rt.

File: hippiestation/code/modules/mob/living/carbon/human/say.dm [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/mob/living/carbon/human/proc/forcesay_allowcomms(append)
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/mob/living/carbon/human/say.dm:1]: wat

Comment (ghost) [reply hippiestation/code/modules/mob/living/carbon/human/say.dm:1]: you know if you get hit or stunned while typing (unless you use hotkey), you'll say the message with like"-ACK!"

this proc is basically a copy that doesn't remove the `;` or `:` from the message when doing so

**Comment by ghost** [comment]
#### Test Merge Removed

**Comment by ghost** [comment]
#### Test Merge Deployed

##### Server Instance
HippieStation

##### Revision
Origin: 7b5f3102b0b1a98563759da88f2bbe079b2a0d34
Pull Request: 2e04aded65eb8ac6b7a8d988acd5aab83871248c
Server: 63ddf60ff3c7af7d07cf178b5149dbcd27161c43

##### Comment
stupid vivadel

**Comment by nemvar** [comment]
Upstream fag here. Reminder that /tg/ has an improved version of guardian bombs. https://github.com/tgstation/tgstation/pull/45111 The summoner can interact with them as if they were normal objects and it doesn't glitch out objects that heavily use overlays. Port at your leasure.

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 4859b9b3103432e2578263f3f6ded68784272451
Pull Request: 2e04aded65eb8ac6b7a8d988acd5aab83871248c
Server: dbc6bcaaf389ef64cb5d01cc8eb55255815cdff8

##### Comment
stupid vivadel

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: f16df96751028237558ca7be8823c327360d0d49
Pull Request: 5f0d0ddfcff4d095cf272af8a650ef296b74ac49
Server: 89cf6dfa5ffb75c8069fb0a88be306f2b1a76f93

##### Comment
stupid vivadel

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 2b653e52baa133e569969313d4feb821bbd72739
Pull Request: f49110b5ced75251bcdb651cdf725036461bfbc0
Server: 77493f9358995cd8bfb13cc3565f0f6aedc91b37

##### Comment
is this done yet

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 2b653e52baa133e569969313d4feb821bbd72739
Pull Request: a82c7b517812e64c40eec61b442aa5fbfff23b25
Server: 7f2fffa0c30109ad53aef33e42a1e321acb99c1a

##### Comment
Notorious B.I.G.

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 8a5e687aaa1f2fb51bd9371839a5c55a73628a95
Pull Request: 280c0553c33ddcab05f5aac789dda9f3a80dcc90
Server: 685f455564f1b9d1445abe09c63c62fee0b5ae2d

##### Comment
UGH

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 8a5e687aaa1f2fb51bd9371839a5c55a73628a95
Pull Request: ae829c56150137aba48a08fd171706cca1c1f437
Server: 55cf0f74e63ea2a4513b0f3c2885bf1b06beabcb

##### Comment
ash storm = free arrows!

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: c70377967d5ce657b28a027b13c1691626873636
Pull Request: a4fa8d43591a5e5c9369ef0f6cf8d99ac45582b1
Server: 4489814bc0bf485751d085d69a2152b8b2f3be1b

##### Comment
ash storm = free arrows!

**Comment by ghost** [comment]
#### Test Merge Updated

##### Server Instance
HippieStation

##### Revision
Origin: 7d245756cf0be67f3f92e812dcac77cbe68cfd6c
Pull Request: a4fa8d43591a5e5c9369ef0f6cf8d99ac45582b1
Server: c75fffb5e43eda8dd3c6ffafdde0fb17d73ae993

##### Comment
ash storm = free arrows!

**Comment by ghost** [comment]
#### Test Merge Removed

**Comment by Manly-Dwarf** [comment]
It's about time we got this merged already.

**Review by Manly-Dwarf — CHANGES_REQUESTED** [review]
Replace every instance of Stand with "guardian spirit." I get that sounds like total cancer, but it should not be such an obvious reference.

File: code/__DEFINES/~hippie_defines/traits.dm [hunk 2]
```diff
@@ -9,4 +10,5 @@
 #define LAG_STONE_TRAIT "lag_stone_trait"
 #define GHOST_STONE_TRAIT "ghost_stone_trait"
 #define GAUNTLET_TRAIT "gauntlet_trait"
+#define STAND_TRAIT	"stand_trait"
```
Comment (Manly-Dwarf) [inline code/__DEFINES/~hippie_defines/traits.dm]: GUARDIAN_TRAIT

File: code/game/gamemodes/meteor/meteors.dm [hunk 3]
```diff
@@ -4,13 +4,13 @@ GLOBAL_VAR_INIT(meteor_wave_delay, 625) //minimum wait between waves in tenths o
...
-						  /obj/effect/meteor/flaming=1, /obj/effect/meteor/irradiated=3)) //for normal meteor event
+						  /obj/effect/meteor/flaming=1, /obj/effect/meteor/irradiated=3, /obj/effect/meteor/stand = 0.05)) //for normal meteor event | hippie -- stand arrow
 
 GLOBAL_LIST_INIT(meteors_threatening, list(/obj/effect/meteor/medium=4, /obj/effect/meteor/big=8, \
-						  /obj/effect/meteor/flaming=3, /obj/effect/meteor/irradiated=3)) //for threatening meteor event
+						  /obj/effect/meteor/flaming=3, /obj/effect/meteor/irradiated=3, /obj/effect/meteor/stand = 0.05)) //for threatening meteor event | hippie -- stand arrow
 
 GLOBAL_LIST_INIT(meteors_catastrophic, list(/obj/effect/meteor/medium=5, /obj/effect/meteor/big=75, \
-						  /obj/effect/meteor/flaming=10, /obj/effect/meteor/irradiated=10, /obj/effect/meteor/tunguska = 1)) //for catastrophic meteor event
+						  /obj/effect/meteor/flaming=10, /obj/effect/meteor/irradiated=10, /obj/effect/meteor/tunguska = 1, /obj/effect/meteor/stand = 0.075)) //for catastrophic meteor event | hippie -- stand arrow
```
Comment (Manly-Dwarf) [inline code/game/gamemodes/meteor/meteors.dm]: /obj/effect/meteor/guardian or /obj/effect/meteor/spirit

File: hippiestation/code/datums/ruins/lavaland.dm [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/datum/map_template/ruin/lavaland/stand_arrow
+	name = "Nijimura's Attic"
```
Comment (Manly-Dwarf) [inline hippiestation/code/datums/ruins/lavaland.dm]: bro

File: hippiestation/code/game/gamemodes/meteor/meteors.dm [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/obj/effect/meteor/stand
```
Comment (Manly-Dwarf) [inline hippiestation/code/game/gamemodes/meteor/meteors.dm]: See above

File: hippiestation/code/modules/guardian/abilities/_ability.dm [hunk 6]
```diff
@@ -0,0 +1,46 @@
+/obj/effect/proc_holder/spell/targeted/stand
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/_ability.dm]: all of this

File: hippiestation/code/modules/guardian/abilities/major/assassin.dm [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/datum/guardian_ability/major/assassin
+	name = "Assassin"
+	desc = "The stand can sneak up on people and do a powerful attack."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/major/assassin.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/major/gravity.dm [hunk 8]
```diff
@@ -0,0 +1,52 @@
+/datum/guardian_ability/major/gravity
+	name = "Gravity"
+	desc = "The stand's punches apply heavy gravity to whatever it punches."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/major/gravity.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/major/healing.dm [hunk 9]
```diff
@@ -0,0 +1,45 @@
+/datum/guardian_ability/major/healing
+	name = "Healing"
+	desc = "Allows the stand to heal anything, living or inanimate, by touch."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/major/healing.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/major/scout.dm [hunk 10]
```diff
@@ -0,0 +1,36 @@
+/datum/guardian_ability/major/scout
+	name = "Scout"
+	desc = "The stand can turn near-invisible and invincible and scout the station, although it cannot attack anything."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/major/scout.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/major/time.dm [hunk 11]
```diff
@@ -0,0 +1,89 @@
+/datum/guardian_ability/major/time
+	name = "Time Erasure"
+	desc = "The stand can erase a short period of time."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/major/time.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/minor/snares.dm [hunk 12]
```diff
@@ -0,0 +1,57 @@
+/datum/guardian_ability/minor/snare
+	name = "Surveillance Snares"
+	desc = "The stand can lay a surveillance snare, which alerts the guardian and the user to anyone who crosses it."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/minor/snares.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/minor/teleport.dm [hunk 13]
```diff
@@ -0,0 +1,99 @@
+/datum/guardian_ability/minor/teleport
+	name = "Teleportation Pad"
+	desc = "The stand can prepare a teleportation pad, and teleport things to it afterwards."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/minor/teleport.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/minor/teleport.dm [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+/datum/guardian_ability/minor/teleport/Apply()
+	..()
+	guardian.verbs += /mob/living/simple_animal/hostile/guardian/proc/Beacon
+
+/datum/guardian_ability/minor/teleport/Remove()
+	..()
+	guardian.verbs -= /mob/living/simple_animal/hostile/guardian/proc/Beacon
+
+/obj/effect/proc_holder/spell/targeted/stand/teleport
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/minor/teleport.dm]: Stand is here, again.

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 15]
```diff
@@ -0,0 +1,174 @@
+/datum/guardian_ability/major/special/onewayroad // a cookie for you if you get the reference
+	name = "Absolution"
+	desc = "The stand forms an absolute shield around it's user, protecting them from all harm."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+/datum/guardian_ability/major/special/onewayroad/Manifest()
+	update_status()
+
+/datum/guardian_ability/major/special/onewayroad/Recall()
+	update_status()
+
+/datum/guardian_ability/major/special/onewayroad/Berserk()
+	if(!HAS_TRAIT(guardian, TRAIT_ONEWAYROAD))
+		ADD_TRAIT(guardian, TRAIT_ONEWAYROAD, STAND_TRAIT)
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: GUARDIAN_TRAIT

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+		for(var/obj/effect/particle_effect/foam/F in T)
+			F.visible_message("<span class='danger'>\The [F] is dispersed into a million tiny particles!</span>")
+			qdel(F)
+
+/datum/guardian_ability/major/special/onewayroad/proc/update_status()
+	if(!guardian || !guardian.summoner)
+		return
+	if(guardian.loc == guardian.summoner)
+		if(HAS_TRAIT(guardian.summoner, TRAIT_ONEWAYROAD))
+			REMOVE_TRAIT(guardian.summoner, TRAIT_ONEWAYROAD, STAND_TRAIT)
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: GUARDIAN_TRAIT

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+		if(HAS_TRAIT(guardian.summoner, TRAIT_NOBREATH))
+			REMOVE_TRAIT(guardian.summoner, TRAIT_NOBREATH, STAND_TRAIT)
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: GUARDIAN_TRAIT

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+	else
+		ADD_TRAIT(guardian.summoner, TRAIT_ONEWAYROAD, STAND_TRAIT)
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: GUARDIAN_TRAIT

File: hippiestation/code/modules/guardian/abilities/special/absolution.dm [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+		ADD_TRAIT(guardian.summoner, TRAIT_NOBREATH, STAND_TRAIT) // this kinda simulates the "Absolution filters out harmful gases around the user" thing better than constantly parsing gas strings
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/absolution.dm]: GUARDIAN_TRAIT

File: hippiestation/code/modules/guardian/abilities/special/pocket.dm [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+	name = "??? INVALID COORDINATES ???"
+	dynamic_lighting = DYNAMIC_LIGHTING_DISABLED
+	has_gravity = STANDARD_GRAVITY
+	noteleport = TRUE
+	unique = FALSE
+	requires_power = FALSE
+
+/datum/guardian_ability/major/special/pocket
+	name = "Pocket Dimension"
+	desc = "The stand can access a small pocket dimension, bringing it's owner with it as well."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/pocket.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/special/pocket.dm [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+	if(!LAZYLEN(zs))
+		return
+	var/pocket = zs[1]
+	if(L.z != pocket)
+		take_effects(L)
+
+/obj/effect/proc_holder/spell/self/pocket_dim/proc/add_effects(mob/living/L)
+	L.status_flags |= GODMODE
+	RegisterSignal(L, COMSIG_MOVABLE_MOVED, .proc/check_if_teleport)
+	for(var/mob/living/simple_animal/hostile/guardian/stando in L.hasparasites())
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/pocket.dm]: Less on the nose, and code only.

File: hippiestation/code/modules/guardian/abilities/special/time.dm [hunk 23]
```diff
@@ -0,0 +1,19 @@
+/datum/guardian_ability/major/special/timestop
+	name = "Time Stop"
+	desc = "The stand can stop time in a localized area."
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/time.dm]: The guardian/spirit

File: hippiestation/code/modules/guardian/abilities/special/time.dm [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+	cost = 5
+	spell_type = /obj/effect/proc_holder/spell/aoe_turf/conjure/timestop/stand
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/time.dm]: guardian

File: hippiestation/code/modules/guardian/abilities/special/time.dm [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+/datum/guardian_ability/major/special/timestop/Berserk()
+	guardian.RemoveSpell(spell)
+	spell = new /obj/effect/proc_holder/spell/aoe_turf/conjure/timestop/stand/berserk
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/time.dm]: guardian

File: hippiestation/code/modules/guardian/abilities/special/time.dm [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+	guardian.AddSpell(spell)
+
+/obj/effect/proc_holder/spell/aoe_turf/conjure/timestop/stand
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/time.dm]: guardian

File: hippiestation/code/modules/guardian/abilities/special/time.dm [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+	invocation_type = "none"
+	clothes_req = FALSE
+	staff_req = FALSE
+	summon_type = list(/obj/effect/timestop)
+
+/obj/effect/proc_holder/spell/aoe_turf/conjure/timestop/stand/berserk
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/abilities/special/time.dm]: guardian

File: hippiestation/code/modules/guardian/guardianbuilder.dm [hunk 28]
```diff
@@ -0,0 +1,288 @@
...
+		var/mob/dead/observer/C = pick(candidates)
+		var/mob/living/simple_animal/hostile/guardian/G = new(user, theme)
+		if(guardian_name)
+			G.real_name = guardian_name
+			G.name = guardian_name
+			G.custom_name = TRUE
+		G.summoner = user
+		G.key = C.key
+		G.mind.enslave_mind_to_creator(user)
+		var/datum/antagonist/stand/S = new
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/guardianbuilder.dm]: guardian

File: hippiestation/code/modules/guardian/guardiandatum.dm [hunk 29]
```diff
@@ -0,0 +1,28 @@
+// why does this exist? simply so stands show up on roundend screen.
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/guardiandatum.dm]: All of this.

File: hippiestation/code/modules/guardian/standarrow.dm [hunk 30]
```diff
@@ -0,0 +1,190 @@
+/obj/item/stand_arrow
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/guardian/standarrow.dm:1]: This is fine, mostly, as it's "mysterious arrow."

File: hippiestation/code/modules/research/xenobiology/crossbreeding/_misc.dm [hunk 31]
```diff
@@ -0,0 +1,5 @@
+/datum/component/dejavu
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/research/xenobiology/crossbreeding/_misc.dm]: I don't think this is used at all, is it?

**Review by Manly-Dwarf — DISMISSED** [review]

**Comment by AnthonyJNgo** [comment]
Obvious references in the code don't really matter to be honest, the only big thing would be if it were obvious references in the game.

**Review by yoyobatty — APPROVED** [review]

**Pull request merged**